include/deal.II/base/quadrature_point_data.h

```diff
@@ -0,0 +1,808 @@
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii__quadrature_point_data_h
+#define dealii__quadrature_point_data_h
+
+#ifdef DEAL_II_WITH_CXX11
+
+#include <deal.II/base/quadrature.h>
+#include <deal.II/base/config.h>
+#include <deal.II/base/subscriptor.h>
+#include <deal.II/grid/tria.h>
+#include <deal.II/grid/tria_accessor.h>
+#include <deal.II/distributed/tria.h>
+#include <deal.II/fe/fe_tools.h>
+#include <deal.II/fe/fe.h>
+#include <deal.II/lac/vector.h>
+
+#include <vector>
+#include <map>
+#include <typeinfo>
+
+DEAL_II_NAMESPACE_OPEN
+
+/*!@addtogroup Quadrature */
+/*@{*/
+
+/**
+ * A class for storing at each cell @p CellIteratorType a vector of data @p DataType .
+ * The underlying structure and the initialize() method of this class is designed
+ * in such a way that one could use different child classes derived from the
+ * base DataType class to store data on a given cell. This implies the usage of pointers,
+ * in our case -- std::shared_ptr().
+ *
+ * @note Note that the data type stored on each cell can be different.
+ * However, within the cell this class stores a vector of objects of a single data type.
+ * For this reason, this class may not be sufficiently flexible when, for example,
+ * adopting a level-set approach to describe material behavior.
+ *
+ * @author Denis Davydov, Jean-Paul Pelteret, 2016
+ */
+template <typename CellIteratorType, typename DataType>
+class CellDataStorage
+{
+public:
+  /**
+   * A typedef for a vector of shared pointers with arbitrary data type @p T .
+   */
+  template <typename T> using CellVector = std::vector<std::shared_ptr<T> >;
+
+  /**
+   * Default constructor.
+   */
+  CellDataStorage();
+
+  /**
+   * Default virtual destructor.
+   */
+  virtual ~CellDataStorage();
+
+  /**
+   * Initialize data on the @p cell to store @p number_of_data_points of objects of type @p T .
+   * The typename @p T is possibly another class which is derived from the
+   * base @p DataType class. In order to initialize the vector of objects
+   * we have to assume that the class @p T has a default constructor.
+   * This function has to be called on every cell where data is to be stored.
+   *
+   * After the data is initialized, it can be modified using the get_data()
+   * function.
+   *
+   * @note Subsequent calls of this function with the same @p cell will not
+   * alter the objects associated with it. In order to remove the stored data,
+   * use the erase() function.
+   *
+   * @note It is possible to use different types @p T for different cells which
+   * may reflect, for example, different constitutive models of continuum mechanics
+   * in different parts of the domain.
+   */
+  template<typename T=DataType>
+  void initialize(const CellIteratorType &cell,
+                  const unsigned int number_of_data_points);
```

Comment:
Maybe add
```
  @pre The type @p T needs to either equal @p DataType, or be a class derived from @p DataType.
```
This could also be stated as part of a `@tparam` statement.

I wonder whether it's worth using the C++11 facilities in the implementation of the function to figure out whether this precondition is indeed satisfied. There must be something like `std::is_derived_from<T,U>::value` or similar that could be used in a `static_assert`.